### PR TITLE
Kp/classification exp

### PR DESCRIPTION
**Pull request by kprokofi**
Updated loading weight, optimizers, and schedulers for aux model. Removed extra-weights params for consistency with main model. Now, loading weights carried out from config only.

@LeonidBeynenson FYI

### CODE REVIEW

File: tools/main.py [hunk 1]
```diff
@@ -24,31 +25,42 @@ def build_datamanager(cfg, classification_classes_filter=None):
...
-    if cfg.use_gpu:
+    if aux_cfg.use_gpu:
         assert device_ids is not None
 
         model = DataParallel(model, device_ids=device_ids, output_device=0).cuda(device_ids[0])
+    if lr is not None:
+        aux_cfg.train.lr = lr
+        print(f"setting learning rate from main model, estimated by lr finder: {lr}")
+    if aux_cfg.loss.name == 'am_softmax':
+        s = compute_s(num_classes[0])
```
Comment (sovrasov) [inline tools/main.py]: I think we shouldn't hack config right in builder and instead introduce a parameter in config.

File: tools/main.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 from scripts.script_utils import build_base_argparser, reset_config, check_classes_consistency
 
 import torchreid
+import numpy as np
```
Comment (sovrasov) [inline tools/main.py]: Seems like numpy is not needed anymore

File: tools/main.py [hunk 3]
```diff
@@ -82,9 +89,6 @@ def main():
     num_params, flops = compute_model_complexity(model, (1, 3, cfg.data.height, cfg.data.width))
     print('Main model complexity: params={:,} flops={:,}'.format(num_params, flops))
 
-    if cfg.model.load_weights and check_isfile(cfg.model.load_weights):
```
Comment (sovrasov) [inline tools/main.py:85]: This should be restored

Comment (kprokofi) [reply tools/main.py:85]: it's done on 150 line

File: tools/main.py [hunk 4]
```diff
@@ -140,9 +144,13 @@ def main():
...
     if cfg.model.resume and check_isfile(cfg.model.resume):
         cfg.train.start_epoch = resume_from_checkpoint(
-            cfg.model.resume, model, optimizer=optimizer, scheduler=scheduler
-        )
+            cfg.model.resume, model, optimizer=optimizer,scheduler=scheduler
+            )
 
+    elif cfg.model.load_weights and check_isfile(cfg.model.load_weights):
+        load_pretrained_weights(model, cfg.model.load_weights)
+
```
Comment (LeonidBeynenson) [inline tools/main.py]: Kirill, is it possible to move these lines higher -- where `load_pretrained_weights` were earlier?
Is it possible to make loading pretrained weights or resuming _before_  `DataParallel` wrapping `model = DataParallel(model, ...)`

Why as about that:
* I think it is better load weights before DataParallel-related actions
* for NNCF wrapping it is more convenient

If from your point of view it is a wrong idea to move this block of code higher (may be due to some details concerning `resume`), I will make a workaround for NNCF, but, nevertheless, I think it is better to move it (if it is possible).

Comment (kprokofi) [reply tools/main.py]: done

File: torchreid/engine/image/am_softmax.py [hunk 5]
```diff
@@ -47,13 +47,17 @@ class ImageAMSoftmaxEngine(Engine):
     def __init__(self, datamanager, model, optimizer, reg_cfg, metric_cfg, scheduler=None, use_gpu=False, save_chkpt=True,
                  train_patience=10, early_stoping = False, lb_lr = 1e-5, softmax_type='stock', label_smooth=False,
                  margin_type='cos', epsilon=0.1, aug_type=None, decay_power=3, alpha=1., size=(224, 224), max_soft=0.0,
-                 reformulate=False, aug_prob=1., conf_penalty=False, pr_product=False, m=0.35, s=10, end_s=None, duration_s=None,
+                 reformulate=False, aug_prob=1., conf_penalty=False, pr_product=False, m=0.35, compute_s=False, s=10, end_s=None, duration_s=None,
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py]: Where is this parameter set in engine and config?

File: torchreid/engine/image/am_softmax.py [hunk 6]
```diff
@@ -47,13 +47,17 @@ class ImageAMSoftmaxEngine(Engine):
...
+                 reformulate=False, aug_prob=1., conf_penalty=False, pr_product=False, m=0.35, s=10, compute_s=False, end_s=None,
+                 duration_s=None, skip_steps_s=None, enable_masks=False, adaptive_margins=False, class_weighting=False,
+                 attr_cfg=None, base_num_classes=-1, symmetric_ce=False, mix_weight=1.0, enable_rsc=False, enable_sam=False):
         super(ImageAMSoftmaxEngine, self).__init__(datamanager, model, optimizer, scheduler, use_gpu, save_chkpt,
                                                     train_patience, lb_lr, early_stoping)
 
         assert softmax_type in ['stock', 'am']
+        if compute_s and softmax_type == 'am':
+            print(f"computed margin scale for dataset: {s}")
+            s = self.compute_s(num_class[0])
```
Comment (LeonidBeynenson) [inline torchreid/engine/image/am_softmax.py]:
```suggestion
            s = self.compute_s(num_class[0])
            print(f"computed margin scale for dataset: {s}")
```

Comment (LeonidBeynenson) [reply torchreid/engine/image/am_softmax.py]: (Changed the order of lines to prnt the calculated value of `s`

**Review by LeonidBeynenson — CHANGES_REQUESTED** [review]
Kirill, one more thing that I would like to ask you:
Could you return back the script `run_classification.py`?

The reasoning:
1. I think it may be important for any user of the repository to understand how the trainings were started and which options were used.
  (I would prefer not to hide this from the users)
2. And (unfortunately) I have to say that there is probability that in the future retrain of the models will be required again (new pytorch, other new libraries, etc).

I really think the script should be returned and moved to some subfolder like `tools/auxiliary/`

I would be glad if you return the script "as is".
(And will be twice glad if change it a bit -- add some options to the script like `--data_root` (is not used now), etc -- but IMHO this is optional.)

**Comment by kprokofi** [comment]
@LeonidBeynenson Are there any other comments/proposals or merge?

**Review by LeonidBeynenson — APPROVED** [review]

**Comment by LeonidBeynenson** [comment]
> @LeonidBeynenson Are there any other comments/proposals or merge?

No, it seems now everythink is OK. Thank you.

**Pull request merged**